modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/queue/KafkaConsumer.scala

```diff
@@ -1,56 +1,134 @@
 package ch.epfl.bluebrain.nexus.kg.service.queue
 
 import akka.Done
-import akka.actor.ActorSystem
+import akka.actor._
 import akka.kafka.ConsumerMessage.CommittableOffsetBatch
 import akka.kafka.scaladsl.Consumer
 import akka.kafka.{ConsumerSettings, Subscriptions}
+import akka.pattern.BackoffSupervisor.{CurrentChild, GetCurrentChild}
+import akka.pattern.{Backoff, BackoffSupervisor, ask}
 import akka.stream.ActorMaterializer
 import akka.stream.scaladsl.Sink
-import io.circe.Decoder
+import akka.util.Timeout
+import com.google.common.annotations.VisibleForTesting
+import io.circe._
 import io.circe.parser._
 import journal.Logger
 
+import scala.concurrent.duration._
 import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
 
-object KafkaConsumer {
+class KafkaConsumerWrapper[T](settings: ConsumerSettings[String, String],
+                       index: T => Future[Unit],
+                       topic: String,
+                       decoder: Decoder[T])
+    extends Actor {
+
+  private implicit val ec: ExecutionContext  = context.dispatcher
+  private implicit val to: Timeout           = 30.seconds
+  private implicit val mt: ActorMaterializer = ActorMaterializer()
+  private implicit val ed: Decoder[T]        = decoder
 
   private val log = Logger[this.type]
 
-  /**
-    * Starts a Kafka consumer that reads messages from a particular ''topic'', decodes them into
-    * events of type ''T'' and indexes them using the provided function.
-    *
-    * @param settings an instance of [[akka.kafka.ConsumerSettings]]
-    * @param index    the indexing function that is applied to received events
-    * @param topic    the Kafka topic to read messages from
-    * @param as       an implicitly available actor system
-    * @param D        an implicitly available Circe decoder to deserialize received messages
-    * @tparam T       the event type
-    */
-  final def start[T](settings: ConsumerSettings[String, String], index: T => Future[Unit], topic: String)(
-      implicit as: ActorSystem,
-      D: Decoder[T]): Future[Done] = {
-    implicit val ec: ExecutionContext  = as.dispatcher
-    implicit val mt: ActorMaterializer = ActorMaterializer()
+  override def receive: Receive = {
+    case value: String =>
+      decode[T](value) match {
+        case Right(event) =>
+          log.debug(s"Received message: $value")
+          index(event).onComplete {
+            case Success(_) => ()
+            case Failure(e) => log.error(s"Failed to index event: $event; skipping.", e)
+          }
+          sender ! Done
```

Comment:
I think `Done` should only be sent once the indexing function completes otherwise there's no backpressure and since the mailbox is unbounded it can result and out of memory errors.